lib/spack/llnl/util/tty/log.py

```diff
@@ -445,62 +565,58 @@ def _writer_daemon(self, stdin):
 
         log_file = self.log_file
 
-        def handle_write(force_echo):
-            # Handle output from the with block process.
-            # If we arrive here it means that in_pipe was
-            # ready for reading : it should never happen that
-            # line is false-ish
-            line = in_pipe.readline()
-            if not line:
-                return (True, force_echo)  # break while loop
-
-            # find control characters and strip them.
-            controls = control.findall(line)
-            line = re.sub(control, '', line)
-
-            # Echo to stdout if requested or forced
-            if echo or force_echo:
-                try:
-                    if termios:
-                        conf = termios.tcgetattr(sys.stdout)
-                        tostop = conf[3] & termios.TOSTOP
-                    else:
-                        tostop = True
-                except Exception:
-                    tostop = True
-                if not (tostop and _is_background_tty()):
-                    sys.stdout.write(line)
-                    sys.stdout.flush()
-
-            # Stripped output to log file.
-            log_file.write(_strip(line))
-            log_file.flush()
-
-            if xon in controls:
-                force_echo = True
-            if xoff in controls:
-                force_echo = False
-            return (False, force_echo)
-
         try:
-            with _keyboard_input(stdin):
+            with keyboard_input(stdin):
                 while True:
                     # No need to set any timeout for select.select
                     # Wait until a key press or an event on in_pipe.
-                    rlist, _, _ = select.select(istreams, [], [])
+                    try:
+                        rlist, _, _ = select.select(istreams, [], [])
+                    except select.error as e:
+                        # This happens we get a signal while in select()
+                        if e.args[0] == errno.EINTR:
+                            continue
+                        raise
+
                     # Allow user to toggle echo with 'v' key.
                     # Currently ignores other chars.
                     # only read stdin if we're in the foreground
-                    if stdin in rlist and not _is_background_tty():
-                        if stdin.read(1) == 'v':
-                            echo = not echo
+                    if stdin in rlist and not _is_background_tty(stdin):
+                        # it's possible to be backgrounded between the above
+                        # check and the read, so we ignore SIGTTIN here.
+                        with ignore_signal(signal.SIGTTIN):
+                            try:
+                                if stdin.read(1) == 'v':
+                                    echo = not echo
+                            except IOError as e:
+                                # if SIGTTIN is ignored,
```

Comment:
yeah I forgot to finish this comment. The protocol is that a read in the bg will generate `SIGTTIN`, and if you ignore `SIGTTIN`, the call returns 1 with `EIO` to let you know it failed (instead of the signal).  We